src/D3MHub.sol

```diff
@@ -221,115 +221,64 @@ contract D3MHub {
+                _pool.withdraw(amount);
+                daiJoin.join(address(this), amount);
 
-            if (amount + fees > availableAssets) {
-                unchecked {
-                    fees = availableAssets - amount;
-                }
+                // assetsBalance == art tracked in vat (as _fix is called previously) => amount <= art
+                vat.frob(ilk, address(_pool), address(_pool), address(this), -int256(amount), -int256(amount));
+                vat.slip(ilk, address(_pool), -int256(amount));
             }
-        }
+        } else if (mode == Mode.D3M_CULLED || mode == Mode.MCD_CAGED) {
+            if (amount > 0) {
+                address urn = mode == Mode.D3M_CULLED ? address(_pool) : address(end);
 
-        if (amount == 0 && fees == 0) {
-            emit Unwind(ilk, 0, 0);
-            return;
-        }
-
-        // To save gas you can bring the fees back with the unwind
-        uint256 total = amount + fees;
-        _pool.withdraw(total);
-        daiJoin.join(address(this), total);
+                uint256 toSlip = _min(vat.gem(ilk, urn), amount);
+                require(toSlip <= MAXINT256, "D3MHub/overflow");
 
-        // normalized debt == erc20 DAI to pool (Vat rate for D3M ilks fixed to 1 RAY)
+                _pool.withdraw(amount);
+                daiJoin.join(address(this), amount);
 
-        if (mode == Mode.NORMAL) {
-            vat.frob(ilk, address(_pool), address(_pool), address(this), -int256(amount), -int256(amount));
-            vat.slip(ilk, address(_pool), -int256(amount));
-            vat.move(address(this), vow, fees * RAY);
-        } else if (mode == Mode.D3M_CULLED) {
-            vat.slip(ilk, address(_pool), -int256(amount));
-            vat.move(address(this), vow, total * RAY);
-        } else {
-            // This can be done with the assumption that the price of 1 collateral unit equals 1 DAI.
-            // That way we know that the prev End.skim call kept its gap[ilk] emptied as the CDP was always collateralized.
-            // Otherwise we couldn't just simply take away the collateral from the End module as the next line will be doing.
-            vat.slip(ilk, address(_end), -int256(amount));
-            vat.move(address(this), vow, total * RAY);
-        }
+                vat.slip(ilk, urn, -int256(toSlip));
+                vat.move(address(this), vow, amount * RAY);
+            }
+        } else revert("D3MHub/unknown-mode");
 
-        emit Unwind(ilk, amount, fees);
+        emit Unwind(ilk, amount);
     }
 
-    function _fix(bytes32 ilk, address _pool) internal returns (uint256 diff) {
+    function _fix(bytes32 ilk, address _pool, uint256 assets) internal returns (uint256 diff) {
         (uint256 ink, uint256 art) = vat.urns(ilk, _pool);
+        if (assets > ink) {
+            uint256 _diff = assets - ink;
+            require(_diff <= MAXINT256, "D3MHub/overflow");
+            vat.slip(ilk, address(_pool), int256(_diff));
```

Comment:
`_pool` doesn't need casting to address.

Reply:
Same here, this should be solved now.